### PR TITLE
[FSDP/AC] Add test for non-reentrant interleaving

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102692
* #102680
* #102679
* #102672
* __->__ #102669
* #102658
* #102657
* #102655
* #102639
* #102637
* #102508

This is a desired use case and reentrant hits some crashes here, but
seems non-reentrant works well (at least in correctness, memory needs to be
validated on real workloads).

Differential Revision: [D46339161](https://our.internmc.facebook.com/intern/diff/D46339161/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_checkpoint.py [hunk 1]
```diff
@@ -310,6 +312,77 @@ def forward(self, x):
...
+    def forward(self, x):
+        return self.block_module(x)
+
+
+class TestFSDPACInterleaving(FSDPTest):
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_ac_interleave(self):
+        # Works for non-reentrant, fails for reentrant
+        indices = [[0], [0, 1], [0, 1, 2], [1], [1, 2], [2], [0, 2]]
+        for index, use_orig_params in itertools.product(indices, [True, False]):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:361]: will add comments on what this is testing

File: test/distributed/fsdp/test_fsdp_checkpoint.py [hunk 2]
```diff
@@ -310,6 +312,77 @@ def forward(self, x):
         return self.l2(self.relu(self.s2(self.s1(self.l1(x)))))
 
 
+# FSDP(
+# X(
+# A() # no checkpoint
+# Checkpoint(Block())
+# C() # no checkpoint
+# )
+# )
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:321]: nit: Should we indent this properly to show the nesting?

File: test/distributed/fsdp/test_fsdp_checkpoint.py [hunk 3]
```diff
@@ -310,6 +312,77 @@ def forward(self, x):
...
+                    rank=self.rank, checkpoint_indices=index, reentrant=False
+                ).cuda(),
+                use_orig_params=use_orig_params,
+            )
+            for i in range(6):
+                inp = torch.randn(2, 10, device="cuda")
+                model(inp).sum().backward()
+                model_ac(inp).sum().backward()
+                # Ensure gradients are the same
+                with FSDP.summon_full_params([model, model_ac], with_grads=True):
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:379]: nit: `[model, model_ac]` needs to be tuple now based on previous PR.

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_checkpoint.py:379]: Also, I thought that `with_grads=True` only supports `use_orig_params=True`, but it looks like you are taking product over `use_orig_params` in `[True, False]`.

File: torch/distributed/fsdp/flat_param.py [hunk 4]
```diff
@@ -2383,9 +2383,12 @@ def _check_storage_freed(tensor: Tensor):
         )
 
     @staticmethod
-    def _check_storage_allocated(tensor: Tensor):
+    def _check_storage_allocated(tensor: Tensor, fqns: List[str]):
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:2386]: We should fix the type either via `Tuple[str, ...]` or `Iterable[str]` or similar.
```suggestion
    def _check_storage_allocated(tensor: Tensor, fqns: Tuple[str, ...]):
```

File: torch/distributed/fsdp/flat_param.py [hunk 5]
```diff
@@ -2383,9 +2383,12 @@ def _check_storage_freed(tensor: Tensor):
...
         storage_size: int = tensor._typed_storage()._size()
-        _p_assert(storage_size > 0, "Expects storage to be allocated")
+        _p_assert(
+            storage_size > 0,
+            f"Expects storage to be allocated for FlatParam with  original param fqns {fqns}"
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:2390]: nit: Adding a newline before the FQNs generally helps with readability I have found.
```suggestion
            f"Expects storage to be allocated for FlatParameter with original parameter FQNs:\n {fqns}"
```

File: test/distributed/fsdp/test_fsdp_checkpoint.py [hunk 6]
```diff
@@ -310,6 +312,77 @@ def forward(self, x):
...
+        self.lin = nn.Linear(10, 10)
+
+    def forward(self, x):
+        return self.lin(x)
+
+
+class BlockHolder(nn.Module):
+    def __init__(self, rank, checkpoint_indices, reentrant):
+        super().__init__()
+        torch.manual_seed(rank)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:336]: It is not clear to me why we want to initialize the model to be different on each rank.

**Pull request closed without merging**